### PR TITLE
feat(react-viem): upgrade to SDK 2.1.0-alpha.2 + Sepolia + delegation flows (SDK-36)

### PR DESCRIPTION
**Pull request by zama-cremaud**
- Bump @zama-fhe/react-sdk + @zama-fhe/sdk to 2.1.0-alpha.2
- Switch from placeholder addresses to real Sepolia token contracts (USDC Mock + USDT Mock ERC-7984 pairs)
- Replace monolithic page.tsx + inline styles with component architecture
  + globals.css (BalancesCard, ShieldCard, TransferCard, UnshieldCard, PendingUnshieldCard, DelegateDecryptionCard, RevokeDelegationCard, DecryptAsCard)
- Rewrite providers.tsx: ViemSigner with walletKey + refSeededRef pattern, getAddress() EIP-55 normalization, separate IndexedDB instances for storage vs sessionStorage, ZamaSDKEvents.UnshieldPhase1Submitted handler
- Add delegation flows: grant / revoke / decrypt-as with useDelegationStatus live indicator and DelegationNotFoundError / DelegationExpiredError handling
- Add pending unshield persistence via savePendingUnshield + loadPendingUnshield
- Add RelayerWeb proxy with hop-by-hop header filtering and 503 on network failure
- Add src/lib/{config,ethereum,activeUnshield,parseAmount}.ts
- Add client-providers.tsx SSR-safe wrapper (next/dynamic ssr:false)
- Add Playwright e2e tests (22 tests across connect / main / network / delegation)
- Fix NEXT_PUBLIC_SEPOLIA_RPC_URL: use || instead of ?? (Next.js replaces unset NEXT_PUBLIC_* with "" at build time, not undefined)
- Remove typescript: { ignoreBuildErrors: true } from next.config.ts
- Add WALKTHROUGH.md with viem-specific notes (walletKey initial bump, getAddress() normalization, parseAbi() requirement, || vs ??)
- Surface mint error/success feedback in BalancesCard (mintError + mintTxHash)

# Summary

## Scope

- [ ] `@zama-fhe/sdk`
- [ ] `@zama-fhe/react-sdk`
- [ ] `@zama-fhe/test-components`
- [ ] `@zama-fhe/test-nextjs`
- [ ] `@zama-fhe/test-vite`
- [ ] CI/workflows
- [ ] Docs

### CODE REVIEW

**Comment by ghermet** [comment]
## PR Review — Comprehensive Analysis

Reviewed 34 changed files (+2595/−211) using 4 specialized agents: code review, error handling, test coverage, and comment quality.

---

### Critical (must fix before merge)

**1. Open proxy with no path validation — SSRF risk**
`src/app/api/relayer/[...path]/route.ts`

User-supplied path segments are forwarded to `new URL(path.join("/"), RELAYER_URL + "/")` without sanitization. Suggest validating segments against a safe-character pattern or allowlist:

```ts
const SAFE_SEGMENT = /^[a-zA-Z0-9._-]+$/;
for (const segment of path) {
  if (!SAFE_SEGMENT.test(segment)) {
    return new Response(JSON.stringify({ error: "Invalid path" }), {
      status: 400, headers: { "content-type": "application/json" },
    });
  }
}
```

**2. `savePendingUnshield` in onEvent is fire-and-forget — funds-at-risk**
`src/providers.tsx` (onEvent handler)

If IndexedDB write fails (quota exceeded, Safari Private Mode), the pending unshield state is silently lost. User closes tab → cannot resume → funds stuck. Add `.catch()`:

```ts
savePendingUnshield(indexedDBStorage, wrapperAddress, event.txHash)
  .catch((err) => console.error("[Providers] Failed to persist pending unshield:", event.txHash, err));
```

---

### Important (should fix)

**3. USDT approval catch block is overly broad** — `src/components/ShieldCard.tsx`

The catch after `writeContract(approveContract(...))` only re-throws `UserRejectedRequestError` and treats all other errors as "USDT-style, needs reset." Network errors, insufficient gas, nonce issues → 3 failed wallet popups instead of 1 clear error. Narrow the catch to contract reverts:

```ts
if (err instanceof UserRejectedRequestError) throw err;
const isContractRevert = err instanceof Error && 
  (err.message.includes('reverted') || err.name === 'ContractFunctionExecutionError');
if (!isContractRevert) throw err;
needsReset = true;
```

**4. Relayer proxy catch discards all error context** — `src/app/api/relayer/[...path]/route.ts`

Bare `catch {}` returns generic "Relayer unreachable" with zero server-side logging. Add: `catch (err) { console.error("[relayer-proxy]", url.toString(), err); ... }`

**5. `PendingUnshieldCard` silently hides IndexedDB load failures** — `src/components/PendingUnshieldCard.tsx`

If `loadPendingUnshield` fails, the recovery card never renders. User with funds stuck between Phase 1/Phase 2 gets no UI to resume. Surface error state for this funds-at-risk scenario.

**6. Hardcoded `PERMANENT_DELEGATION` sentinel** — `src/components/DecryptAsCard.tsx`

`2n ** 64n - 1n` duplicated from SDK internals. If SDK changes this value, expiry dates silently display wrong. Consider filing an SDK issue to export it as a public constant.

---

### Suggestions (nice to have)

- **`parseAmount.ts`**: Returns `0n` silently on all invalid input — user gets no feedback. Consider returning `{ value, error }` or at minimum `console.warn`.
- **`page.tsx`**: `eslint-disable-line` on `mint.reset` effect — if `mint.reset` is truly stable, include it in deps.
- **`WALKTHROUGH.md`**: Claim "relayer-sdk validates with `getAddress(addr) === addr`" — not confirmed in SDK source. Reword to describe observable behavior.
- **`config.ts`**: Generalize the `||` vs `??` comment to cover all `NEXT_PUBLIC_*` vars.
- **`fixtures.ts`**: Add comment explaining `"0x" + "a".repeat(130)` = 65-byte ECDSA signature.

---

### Test Coverage Gaps

The 22 e2e tests are structural/smoke tests — no transactional flows are exercised:

| Gap | Risk |
|-----|------|
| Shield flow (approval state machine) | Most complex logic, zero coverage |
| Account switch (`accountsChanged`) | No `__emitAccountsChanged` in fixtures |
| Relayer proxy (header filtering, 503) | `_autoMockRelayer` aborts all requests |
| Delegation expiry validation (1h min) | UI validation untested |
| Error state rendering | No alert-error banner tests |

---

### Strengths

- Excellent component decomposition (monolithic page → 8 focused cards)
- Thorough WALKTHROUGH.md documenting every architectural decision
- Well-documented viem-specific gotchas (`||` vs `??`, `parseAbi`, address normalization)
- Solid test infrastructure (`mockWallet`, `mockRpc`, `_autoMockRelayer`)
- Exemplary JSDoc on `activeUnshield.ts`
- Removal of `ignoreBuildErrors: true`

---

*Review performed by 4 parallel agents: code-reviewer, silent-failure-hunter, pr-test-analyzer, comment-analyzer*

File: examples/react-viem/e2e/fixtures.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+/**
+ * Extends the base test to abort all requests to the local /api/relayer proxy for
+ * every test automatically. This prevents real network calls to the Zama relayer
+ * in CI — ZamaProvider handles relayer init failure gracefully, and no test here
+ * exercises actual FHE operations.
+ */
+export const test = base.extend<{ _autoMockRelayer: void }>({
+  _autoMockRelayer: [
+    async ({ page }, use) => {
+      await page.route("**/api/relayer/**", (route) => route.abort());
```
Comment (ghermet) [inline examples/react-viem/e2e/fixtures.ts]: you don't need an auto fixture to do this, you can use the page fixture:
```ts
page:  async ({ page }, use) => {
 await page.route("**/api/relayer/**", (route) => route.abort());
 await use(page)
}
```

File: examples/react-viem/e2e/fixtures.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  }, config);
+}
+
+/**
+ * Intercepts HTTP requests to the Sepolia RPC endpoint and returns minimal
+ * valid JSON-RPC responses. eth_call returns "0x" (empty), which causes
+ * useMetadata to fail gracefully — actionsDisabled stays true and the UI
+ * shows "—" for balances. This keeps tests focused on structure, not data.
+ */
+export async function mockRpc(page: Page) {
```
Comment (ghermet) [inline examples/react-viem/e2e/fixtures.ts]: this function can and probably should be used via the test fixtures:
```ts
mockRpc: async ({ page }), use) =>{
await use(() => mockRpc(page));
}
```
Then in your test deconstruct the fixture 
```ts
test("my test", ({ mockRpc }) => {
 await mockRpc()
[...]
})
```

File: examples/react-viem/e2e/fixtures.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * Key design decisions:
+ * - `eth_accounts` and `eth_requestAccounts` are kept separate so that the SDK
+ *   can call `eth_requestAccounts` internally during initialization without
+ *   affecting what `eth_accounts` returns to page.tsx's auto-detect useEffect.
+ * - `window.__emitChainChanged(chainId)` is exposed so tests can simulate a
+ *   user switching networks in their wallet, triggering the `chainChanged` event
+ *   that page.tsx listens for to transition between screens.
+ */
+export async function mockWallet(page: Page, config: WalletConfig) {
```
Comment (ghermet) [inline examples/react-viem/e2e/fixtures.ts]: this function can and probably should be used via the test fixtures:
```ts
mockWallet: async ({ page }), use) =>{
await use((config: WalletConfig) => mockWallet(page, config));
}
```
Then in your test deconstruct the fixture 
```ts
test("my test", ({ mockWallet }) => {
 await mockWallet({...})
[...]
})
```

File: examples/react-viem/e2e/fixtures.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
+import { test as base, expect, type Page } from "@playwright/test";
+
+export const SEPOLIA_CHAIN_ID_HEX = "0xaa36a7"; // must match src/lib/config.ts
+export const WRONG_CHAIN_ID = "0x1"; // Ethereum mainnet
+export const TEST_ADDRESS = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
```
Comment (ghermet) [inline examples/react-viem/e2e/fixtures.ts]: All these hardcoded values should either be passed as fixtures or declared where they are used.

**Comment by ghermet** [comment]
## Playwright Fixtures Audit

Reviewed all 4 spec files + `fixtures.ts` for proper Playwright fixture usage patterns.

---

### Medium Issues

**1. `mockWallet` and `mockRpc` are plain functions, not Playwright fixtures**

These are exported as `async function(page)` helpers instead of being wired into `test.extend`. This means:
- No automatic setup/teardown — each test must manually call them
- Forgetting `mockRpc` causes tests to silently hit the real Sepolia RPC
- Contrast with `_autoMockRelayer` which IS a proper auto-fixture

`connect.spec.ts` has 3 tests that call neither (intentional for "no wallet" state), but a new contributor adding a test could easily forget `mockRpc` and get flaky network-dependent failures.

**Recommendation:** Convert to parameterized fixtures:
```ts
type TestFixtures = {
  walletConfig: WalletConfig | null;
  mockRpc: void;
  _autoMockRelayer: void;
};

export const test = base.extend<TestFixtures>({
  walletConfig: [null, { option: true }],
  mockRpc: [async ({ page }, use) => {
    await page.route("**/ethereum-sepolia-rpc.publicnode.com**", async (route) => {
      // ... same logic
    });
    await use();
  }, { auto: true }],
  // ...
});

// In tests that need a wallet:
test.use({ walletConfig: { accounts: [TEST_ADDRESS], chainId: SEPOLIA_CHAIN_ID_HEX } });
```

**2. Missing `__emitAccountsChanged` helper**

`__emitChainChanged` exists (used in `network.spec.ts`) but there is no `__emitAccountsChanged` equivalent. The app's `providers.tsx` has a critical `accountsChanged` listener that remounts `ZamaProvider` and bumps `walletKey` — this code path is untestable with the current fixtures.

**Fix:** Add to the `mockWallet` init script:
```ts
(window as any).__emitAccountsChanged = (accounts: string[]) => {
  for (const listener of listeners["accountsChanged"] ?? []) {
    listener(accounts);
  }
};
```

---

### Low Issues

**3. `mockRpc` doesn't guard against `NEXT_PUBLIC_SEPOLIA_RPC_URL` env override**

`mockRpc` intercepts `**/ethereum-sepolia-rpc.publicnode.com**` only. If anyone has `NEXT_PUBLIC_SEPOLIA_RPC_URL` set in `.env.local` (e.g., Infura), the mock won't intercept and tests hit a real RPC. `playwright.config.ts` doesn't clear this env var.

**Fix:** Add to `playwright.config.ts`:
```ts
webServer: {
  command: "npm run dev",
  env: { NEXT_PUBLIC_SEPOLIA_RPC_URL: '' },
  // ...
}
```

**4. CSS class selectors create fragile coupling**

`delegation.spec.ts:28,33` use `page.locator(".card", { hasText: ... })` and `main.spec.ts:49` uses `page.locator(".balance-value")`. These break if CSS is refactored (modules, Tailwind).

**Better:** Use `data-testid` attributes or scope by ARIA roles.

**5. `force: true` click bypasses all actionability checks**

`connect.spec.ts:51` — `click({ force: true })` skips visibility, enabled, and overlay checks. The comment explains the React re-render race, but a safer pattern:
```ts
await page.getByRole("button", { name: /Connect Wallet|Connecting/ }).click();
```

**6. No batch JSON-RPC handling in `mockRpc`**

`postDataJSON()` is parsed as a single object. Viem can send batch requests (array). These would return `{ result: null }`. Not currently exploited, but could cause subtle failures if viem's batching changes.

---

### What's Done Well

- `_autoMockRelayer` is a proper auto-fixture — prevents real relayer calls in every test
- `eth_accounts` / `eth_requestAccounts` separation accurately models real wallet behavior
- `addInitScript` over `evaluate` — correctly injects before page JS runs
- Array spread on returned accounts (`[...cfg.accounts]`) prevents config mutation
- Static `"0x"` for `eth_call` is a deliberate choice keeping tests focused on UI structure

**Comment by ghermet** [comment]
# PR Review — Comprehensive Automated Analysis

4 specialized agents reviewed this PR in parallel: **code quality**, **silent failure hunting**, **test coverage**, and **comment accuracy**.

---

## Critical Issues (3 — must fix before merge)

### 1. Proxy header forwarding leaks cookies to upstream relayer
**`src/app/api/relayer/[...path]/route.ts`** — `forwardHeaders` uses a denylist (hop-by-hop only), so browser `cookie`, `authorization`, and session headers are forwarded to the upstream relayer.

**Fix:** switch to an allowlist (`content-type`, `accept`, `content-length` + the API key).

### 2. Path traversal via `..` segments bypasses `/v2` prefix
**`src/app/api/relayer/[...path]/route.ts`** — `SAFE_SEGMENT = /^[a-zA-Z0-9._-]+$/` allows `..`, which passes validation. `URL` normalization then resolves `["..","admin"]` to `/admin`, escaping the `/v2` path on `RELAYER_URL`.

**Fix:** add `|| segment === '..' || segment === '.'` rejection after the regex check.

### 3. `parseAmount` silently returns `0n` — disabled button with zero user feedback
**`src/lib/parseAmount.ts`** — The bare `catch` swallows all `parseUnits` errors and returns `0n`. User types an invalid amount, button stays disabled with no explanation why. Also has **zero unit tests** despite being called by every financial operation (shield, transfer, unshield).

**Fix:** return a discriminated result (valid/invalid/empty) so callers can show validation messages. Add unit tests for edge cases.

---

## Important Issues (7 — should fix)

### 4. `clearPendingUnshield` is fire-and-forget — phantom "Pending Unshield" cards
**`src/components/UnshieldCard.tsx`** — Called without `await` or `.catch()`. If IndexedDB clear fails, the entry persists and the user sees a stale "Finalize" card for an already-completed unshield.

### 5. `savePendingUnshield` failure = potential stuck tokens with no user notification
**`src/providers.tsx`** — Only logs to `console.error`. If the tab closes after Phase 1, the user has no way to resume the unshield. Surface a visible toast/warning when persistence fails.

### 6. WALKTHROUGH claims `parseUnits` returns `0n` on invalid input — it throws
**`WALKTHROUGH.md` §9** — Attributes the `0n` behavior to viem's `parseUnits`, which actually throws. Developers calling `parseUnits` directly will get uncaught exceptions.

### 7. Comment "`mint.reset` is stable (TanStack Query guarantee)" is incorrect for TQ v5
**`src/app/page.tsx`** — In TanStack Query v5 (confirmed `^5.90.0` in package.json), `useMutation` returns a new object each render; `reset` is not referentially stable. The effect fires every render (harmless/idempotent but the comment is misleading).

### 8. README claims "Webpack" but config doesn't enforce it
**`README.md`** — Says "Webpack — Turbopack not yet supported with WASM" but `next.config.ts` is empty and `next dev` defaults to Turbopack. Either add `--no-turbopack` to the dev script or correct the claim.

### 9. `getChainId` RPC failure has no error handling
**`src/providers.tsx`** — Inside the `RelayerWeb` constructor callback, `ethereum.request({ method: "eth_chainId" })` has no try/catch. Failure propagates as an opaque rejection through RelayerWeb internals.

### 10. Delegation status error discards the actual error message
**`src/components/DecryptAsCard.tsx`** — Shows generic "Unable to check delegation status" without `delegationStatus.error.message`.

---

## Test Coverage Gaps

| Priority | Gap | Effort |
|----------|-----|--------|
| 1 | `parseAmount` unit tests (edge cases, decimals, invalid input) | Low |
| 2 | Relayer proxy path validation + security tests | Medium |
| 3 | `activeUnshield` get/set/clear cycle | Low |
| 4 | E2e test for transaction error display (mock rejection) | Medium |
| 5 | Account switching via `__emitAccountsChanged` | Medium |
| 6 | `delegationErrorMessage` mapping unit tests | Low |
| 7 | ShieldCard USDT approval fallback path | Medium |

The 22 e2e tests are well-structured with a thoughtful mock wallet fixture, but test **UI structure only** — no test exercises an actual transaction flow or error recovery path.

---

## Strengths

- Excellent component decomposition — 8 clean, props-driven components
- Thorough WALKTHROUGH.md with viem-specific gotchas (`||` vs `??`, `getAddress()` normalization)
- Well-designed e2e fixture architecture (`__emitChainChanged`, `__emitAccountsChanged` helpers)
- USDT approval fallback in ShieldCard is correctly implemented and documented
- `providers.tsx` module-level overview comment is exemplary

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by zama-cremaud** [comment]
On latest review, all addressed, except for those:

  3 — parseAmount is intentional, not an error                                                                                                                                                                       
  The code correctly uses our parseAmount wrapper (in src/lib/parseAmount.ts), not viem's parseUnits directly. parseAmount calls parseUnits internally and catches the exception it throws on invalid input, returning 0n instead. The WALKTHROUGH.md was incorrectly attributing the 0n behavior to parseUnits — that has now been corrected to clarify the distinction.                                                                 
                                                            
  5 — No toast on savePendingUnshield failure by design                                                                                                                                                              
 savePendingUnshield is called inside the onEvent handler for ZamaSDKEvents.UnshieldPhase1Submitted, which fires after the Phase 1 transaction has already been submitted on-chain. At that point, the user has already seen their transaction go through. A toast on IndexedDB failure here would be misleading — it would imply the unshield itself failed, when in fact the only thing that failed was the local persistence of  the recovery hash. The .catch() we added logs the error to the console so developers can detect it, but we deliberately avoid surfacing it to the user. If the pending hash is lost, the user would simply need to re-initiate the unshield; the on-chain state is unaffected.                                                                                                                                                         
                                                            
  8 — Webpack note in README is accurate and intentional                                                                                                                                                             
  The README note "Turbopack not yet supported with WASM" reflects a real constraint at the time of writing: Next.js Turbopack does not support the asyncWebAssembly webpack experiment that the Zama FHE worker requires. If this changes in a future Next.js or SDK release, the note should be updated, but removing it now would cause confusion for users who try next dev --turbo and get an opaque WASM error.

**Comment by ankurdotb** [comment]
### Code review

Checked against all issues found in the sister PR #148 (react-ethers). All 10 are correctly addressed — keypairTTL alignment, handleAccountsChanged race, PendingUnshieldCard visibility, ShieldCard try/catch scope, useConfidentialBalance gating, `||` for NEXT_PUBLIC env vars, mint.reset deps, relayer proxy path traversal, .npmrc removal, explicit decrypt pattern.

Found 1 issue:

**WALKTHROUGH.md §6 — `UnshieldPhase1Submitted` timing contradicts code comment** (Low)

The WALKTHROUGH says the event fires "right after Phase 1 is submitted (before mining)" but `providers.tsx` correctly says "after Phase 1 is mined (the SDK awaits the receipt before emitting)." The SDK's `#waitAndFinalizeUnshield` (`token.ts:883`) emits the event after `unwrap()` returns — and `unwrap()` already awaits `waitForTransactionReceipt`. The WALKTHROUGH should match the code.

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by ankurdotb — APPROVED** [review]
